docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+Similarly to loading, both `Vector128` and `Vector256` provide at least five ways of storing them in memory:
+
+```cs
+public static class Vector128
+{
+    public static void Store<T>(this Vector128<T> source, T* destination) where T : unmanaged;
+    public static void StoreAligned<T>(this Vector128<T> source, T* destination) where T : unmanaged;
+    public static void StoreAlignedNonTemporal<T>(this Vector128<T> source, T* destination) where T : unmanaged;
+    public static void StoreUnsafe<T>(this Vector128<T> source, ref T destination) where T : struct;
+    public static void StoreUnsafe<T>(this Vector128<T> source, ref T destination, nuint elementOffset) where T : struct;
+}
+```
+
+For the reasons described for loading, we recommend using the overload that takes managed reference and element offset:
+
+```cs
+public static void StoreUnsafe<T>(this Vector128<T> source, ref T destination, nuint elementOffset) where T : struct;
+```
+
+**Note**: when loading values from one buffer and storing them into another, we need to consider whether they overlap or not. [MemoryExtensions.Overlap](https://learn.microsoft.com/dotnet/api/system.memoryextensions.overlaps#system-memoryextensions-overlaps-1(system-readonlyspan((-0))-system-readonlyspan((-0)))) is an API for doing that.
+
+### Casting
+
+As mentioned before, `Vector128<T>` and `Vector256<T>` are constrained to a specific set of primitive types. `char` is not one of them, but it does not mean that we can't implement vectorized text operations with the new APIs. For primitive types of the same size (and value types that don't contain references), casting is the solution.
+
+[Unsafe.As<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafe.as#system-runtime-compilerservices-unsafe-as-2(-0@)) can be used to get a reference to supported type:
+
+```cs
+void CastingReferences(Span<char> buffer)
+{
+    ref char charSearchSpace = ref MemoryMarshal.GetReference(buffer);
+    ref short searchSpace = ref Unsafe.As<char, short>(ref charSearchSpace);
+    // from now on we can use Vector128<short> or Vector256<short>
+}
+```
+
+Or [MemoryMarshal.Cast<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.memorymarshal.cast#system-runtime-interopservices-memorymarshal-cast-2(system-readonlyspan((-0)))), which casts a span of one primitive type to a span of another primitive type:
+
+```cs
+void CastingSpans(Span<char> chars)
+{
+    Span<short> shorts = MemoryMarshal.Cast<char, short>(chars);
+}
+```
+
+It's also possible to get managed references from unmanaged pointers:
+
+```cs
+void PointerToReference(char* pUtf16Buffer, byte* pAsciiBuffer)
+{
+    // of the same type:
+    ref byte asciiBuffer = ref *pAsciiBuffer;
+    // of different types:
+    ref ushort utf16Buffer = ref *(ushort*)pUtf16Buffer;
+}
+```
+
+We should avoid doing this in the opposite direction, as most engineers will assume that unmanaged pointers are already pinned.
```

Comment:
I'd state this differently: it's only safe to convert a ref to a pointer if it's known that the ref is already pinned.  If it's not, the moment after you get the pointer it could be invalid.